python/ray/experimental/workflow/virtual_actor.py

```diff
@@ -0,0 +1,350 @@
+import inspect
+import logging
+from typing import Optional, List, TYPE_CHECKING, Any, Tuple, Dict
+import uuid
+import weakref
+
+import ray
+from ray.util.inspect import (is_function_or_method, is_class_method,
+                              is_static_method)
+from ray._private import signature
+from ray.experimental.workflow.common import slugify
+from ray.experimental.workflow.storage import Storage, get_global_storage
+from ray.experimental.workflow.recovery import get_latest_output
+
+if TYPE_CHECKING:
+    from ray import ObjectRef
+
+logger = logging.getLogger(__name__)
+
+
+@ray.remote
+def _readonly_method_executor(workflow_id: str, storage: Storage, cls: type,
+                              method_name: str, flattened_args: List):
+    instance = cls.__new__(cls)
+    state = get_latest_output(workflow_id, storage)
+    instance.__setstate__(state)
+    args, kwargs = signature.recover_args(flattened_args)
+    method = getattr(instance, method_name)
+    return method(*args, **kwargs)
+
+
+# Create objects to wrap method invocations. This is done so that we can
+# invoke methods with actor.method.run() instead of actor.method().
+class ActorMethod:
+    """A class used to invoke an actor method.
+
+    Note: This class only keeps a weak ref to the actor, unless it has been
+    passed to a remote function. This avoids delays in GC of the actor.
+
+    Attributes:
+        _actor_ref: A weakref handle to the actor.
+        _method_name: The name of the actor method.
+    """
+
+    def __init__(self, actor, method_name):
+        self._actor_ref = weakref.ref(actor)
```

Comment:
I'm not quite following this one, do you mind share more info or point me to the place where there is more info?

Reply:
I follow this code: https://github.com/ray-project/ray/blob/e7350ff8282660bdd72250c1a553aebb51b20cf4/python/ray/actor.py#L92

Reply:
Is it creating a cyclic ref?